Added MenuTopicManager

Vermunds opened the pull request:
Please double-check if the virtual functions are correct, I'm not 100% sure if I interpreted the vtables correctly. The rest of the class should look like this.
Constructor is at id 34440 and at 34434 for the Dialogue struct.

shadeMe commented on include/RE/Menus/MenuTopicManager.h:
```diff
@@ -0,0 +1,104 @@
+		TESTopicInfo*				 unk30;				  // 30
+		Dialogue*					 currentDialogue;	  // 38
+		CRITICAL_SECTION			 criticalSection;	  // 40
+		RefHandle					 dialogueTarget;	  // 68
```
'target' might be ambiguous w.r.t dialogue. In this context, the ref handle points to the speaker of the response. Therefore, it might be a better idea to rename this member and the next to `dialogueSpeaker` and `lastDialogueSpeaker`.

shadeMe commented on include/RE/Menus/MenuTopicManager.h:
```diff
@@ -0,0 +1,104 @@
+
+
+		// members
+		BSSimpleList<Dialogue*>*	 unk18;				  // 18
```
I have this as [`selectedResponseNode`](https://github.com/TESSkyblivion/Fuz-Ro-D-oh-64/blob/71d10602523ab627d05554fa69be741fc92c538c/FuzRoDohInternals.h#L197).

shadeMe commented on include/RE/Menus/MenuTopicManager.h:
```diff
@@ -0,0 +1,104 @@
+		BSSimpleList<Dialogue*>*	 unk18;				  // 18
+		BSSimpleList<Dialogue*>*	 dialogueList;		  // 20
+		std::uint64_t				 unk28;				  // 28
+		TESTopicInfo*				 unk30;				  // 30
```
I have this as [`rootTopicInfo`](https://github.com/TESSkyblivion/Fuz-Ro-D-oh-64/blob/71d10602523ab627d05554fa69be741fc92c538c/FuzRoDohInternals.h#L200).

shadeMe commented on include/RE/Menus/MenuTopicManager.h:
```diff
@@ -0,0 +1,104 @@
+		BSSimpleList<Dialogue*>*	 dialogueList;		  // 20
+		std::uint64_t				 unk28;				  // 28
+		TESTopicInfo*				 unk30;				  // 30
+		Dialogue*					 currentDialogue;	  // 38
```
If this pointer doesn't get cleared after the dialogue menu is closed, it might be a better idea to name it [`lastSelectedDialogue`](https://github.com/TESSkyblivion/Fuz-Ro-D-oh-64/blob/71d10602523ab627d05554fa69be741fc92c538c/FuzRoDohInternals.h#L201).

shadeMe commented on include/RE/Menus/MenuTopicManager.h:
```diff
@@ -0,0 +1,104 @@
+		std::uint16_t				 padBC;				  // BC
+		BSTArray<TESTopic*>			 unkC0;				  // C0
+	};
+	static_assert(sizeof(MenuTopicManager) == 0xD8);
```
My version of the [class](https://github.com/TESSkyblivion/Fuz-Ro-D-oh-64/blob/71d10602523ab627d05554fa69be741fc92c538c/FuzRoDohInternals.h#L190) has the size at _E0_. IIRC, there's a member at _D8_ that gets referenced in one of the `ActorSoundCallbackManager` methods. Bears investigation.

Vermunds commented:
Thanks for the suggestions.
I also changed the structure of the Dialogue struct according to the version you linked, as apparently I got it wrong.

> My version of the [class](https://github.com/TESSkyblivion/Fuz-Ro-D-oh-64/blob/71d10602523ab627d05554fa69be741fc92c538c/FuzRoDohInternals.h#L190) has the size at _E0_. IIRC, there's a member at _D8_ that gets referenced in one of the `ActorSoundCallbackManager` methods. Bears investigation.

I vaguely remember seeing that somewhere, but I can't seem to find it again. According to the destructor it's size is _D8_ though. I added the last member for now, but no idea if it should be there or not.

Ryan-rsm-McKenzie commented on include/RE/M/MenuTopicManager.h:
```diff
@@ -0,0 +1,116 @@
+
+
+	class MenuTopicManager :
+		public BSTSingletonSDM<MenuTopicManager>,
+		public BSTEventSink<MenuOpenCloseEvent>,
+		public BSTEventSink<PositionPlayerEvent>
```
Can you add offsets to these bases?

Vermunds replied:
Could you take a look at these? `BSTEventSink<PositionPlayerEvent>` seems to be at offset `08` but I can only guess the other two.

Ryan-rsm-McKenzie replied:
```suggestion
		public BSTSingletonSDM<MenuTopicManager>,	// 0x10
		public BSTEventSink<MenuOpenCloseEvent>,	// 0x00
		public BSTEventSink<PositionPlayerEvent>	// 0x08
```

Ryan-rsm-McKenzie commented on include/RE/M/MenuTopicManager.h:
```diff
@@ -0,0 +1,116 @@
+		TESTopicInfo*				 rootTopicInfo;			// 30
+		Dialogue*					 lastSelectedDialogue;	// 38
+		CRITICAL_SECTION			 criticalSection;		// 40
+		RefHandle					 dialogueSpeaker;		// 68
+		RefHandle					 lastDialogueSpeaker;	// 6C - used if the dialogue menu was closed but the NPC is still talking
```
```suggestion
		ObjectRefHandle					 speaker;		// 68
		ObjectRefHandle					 lastSpeaker;	// 6C - used if the dialogue menu was closed but the NPC is still talking
```

Ryan-rsm-McKenzie commented on include/RE/M/MenuTopicManager.h:
```diff
@@ -0,0 +1,116 @@
+		std::uint64_t				 unk28;					// 28
+		TESTopicInfo*				 rootTopicInfo;			// 30
+		Dialogue*					 lastSelectedDialogue;	// 38
+		CRITICAL_SECTION			 criticalSection;		// 40
```
```suggestion
		WinAPI::CRITICAL_SECTION			 criticalSection;		// 40
```

Ryan-rsm-McKenzie commented on include/RE/M/MenuTopicManager.cpp:
```diff
@@ -0,0 +1,11 @@
+
+namespace RE
+{
+	MenuTopicManager* MenuTopicManager::GetSingleton()
+	{
+		REL::Relocation<NiPointer<MenuTopicManager>*> singleton{ REL::ID(514959) };
+		return singleton->get();
+	}
```
Can you inline this in the header? I'd like to avoid compiling additional source files where possible.

shadeMe replied:
Is the singleton instance wrapped in a `NiPointer` though? The NetImmerse smart pointer class is only used with intrusively ref counted classes, IIRC. `MenuTopicManager` isn't one of those, as far as I can tell.

I just have it as a bare pointer in my code.

Ryan-rsm-McKenzie commented on include/RE/M/MenuTopicManager.h:
```diff
@@ -0,0 +1,116 @@
+		bool						 unkBB;					// BB
+		std::uint16_t				 padBC;					// BC
+		BSTArray<TESTopic*>			 unkC0;					// C0
+		std::uint64_t				 unkD8;					// D8 - ???
+	};
+	static_assert(sizeof(MenuTopicManager) == 0xE0);
```
```suggestion
	};
	static_assert(sizeof(MenuTopicManager) == 0xD8);
```

Ryan-rsm-McKenzie commented on include/RE/M/MenuTopicManager.cpp:
```diff
@@ -0,0 +1,11 @@
+{
+	MenuTopicManager* MenuTopicManager::GetSingleton()
+	{
+		REL::Relocation<NiPointer<MenuTopicManager>*> singleton{ REL::ID(514959) };
+		return singleton->get();
```
```suggestion
		REL::Relocation<MenuTopicManager**> singleton{ REL::ID(514959) };
		return *singleton;
```

Ryan-rsm-McKenzie commented on include/RE/M/MenuTopicManager.h:
```diff
@@ -0,0 +1,119 @@
+		struct Dialogue
+		{
+		public:
+			struct ResponseData
+			{
+			public:
+				// members
+				BSString					responseText;	 // 00
+				BGSSceneAction::EmotionType emotionType;	 // 10
+				std::uint32_t				emotionValue;	 // 14
+				BSFixedString				voiceFilePath;	 // 18
+				std::uint64_t				unk20;			 // 20
+				std::uint64_t				unk28;			 // 28
+				std::uint64_t				unk30;			 // 30
+				bool						useEmotionAnim;	 // 38
+				bool						hasLipFile;		 // 39
+				std::uint16_t				pad3A;			 // 3A
+				std::uint32_t				pad3C;			 // 3C
+			};
+			static_assert(sizeof(ResponseData) == 0x40);
```
This is a duplicate of [this](https://github.com/Ryan-rsm-McKenzie/CommonLibSSE/blob/master/include/RE/D/DialogueItem.h#L21-L51).

Vermunds replied:
Oh, didn't saw that one.
